edb/pgsql/metaschema.py

```diff
@@ -2038,14 +2041,23 @@ def __init__(self) -> None:
 class JSONSliceFunction(dbops.Function):
     """Get a JSON array slice."""
     text = r'''
-        SELECT to_jsonb(_slice(
-            (
-                SELECT array_agg(value)
-                FROM jsonb_array_elements(
-                    edgedb.jsonb_assert_type(val, ARRAY['array']))
-            ),
-            start, stop
-        ))
+        SELECT
+            CASE jsonb_typeof(val)
+            WHEN 'array' THEN (
+                to_jsonb(_slice(
+                    (
+                        SELECT array_agg(value)
+                        FROM jsonb_array_elements(val)
+                    ),
+                    start, stop
+                ))
+            )
+            WHEN 'string' THEN (
+                to_jsonb(_slice(val#>>'{}', start, stop))
+            )
+            ELSE
+                edgedb.jsonb_assert_type(val, ARRAY['array'])
```

Comment:
Probably make this an `edgedb.raise`, like for regular indexing?

Reply:
The indexing errors will be raised by the `_slice`, this error is for when the underlying `json` type is not sliceable.

I could inline `edgedb.raise` from `edgedb.jsonb_assert_type`:
```
                edgedb.raise(
                    NULL::jsonb,
                    'wrong_object_type',
                    msg => coalesce(
                        msg,
                        (
                            'expected json '
                            || array_to_string(typenames, ' or ')
                            || '; got json '
                            || coalesce(jsonb_typeof(val), 'UNKNOWN')
                        )
                    ),
                    detail => detail
                )
```

I'm not sure that inlining this would be an improvement.

Reply:
Hm, my actual complaint here is that it will say 'expected json array' when that isn't quite right, but the right fix is just passing in 'string' in the call to jsonb_assert_type.

(It might be worth switching to jsonb_assert_type for JSONIndexByIntFunction and JSONIndexByTextFunction, too, then, but no strong feeling there)

Reply:
+1 to Sully here: in the "ELSE" case just raise an error clearly explaining that the `[]` operator works on JSON arrays and strings, but not on `jsonb_typeof(val)`.

Reply:
OK, I've adjusted the error message to be more consistent with the other slice and index errors for json. I've also added a hint with the information regarding the correct usage.

Speaking of hints, I seem to recall we used to display them in REPL, but I cannot see them for this or for some other errors, like `select <datetime>'1234-11-11';` informing about the expected ISO format. The hints do appear to reach `errormech.py`, though.